### PR TITLE
Implement LWG-3724: `decay-copy` should be constrained

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Towards #2872.

Only one case in the `operator()` of `_All_fn` (the type of `views::all`) needs additional constraints. MSVC STL has already constrained it with `constructible_from`, but sometimes `constructible_from` seems still insufficient.

I'm not sure whether the difference between `constructible_from` and `convertible_to` is observable outside of IFNDR, or whether `weird_string_view` in the test file actually models `view`.

### CODE REVIEW

File: tests/std/tests/P0896R4_views_all/test.cpp [hunk 1]
```diff
@@ -20,6 +20,33 @@ concept CanViewAll = requires(Rng&& r) {
 // Test a silly precomposed range adaptor pipeline
 constexpr auto pipeline = views::all | views::all | views::all | views::all | views::all | views::all;
 
+// Test weird view type with explicit copy constructor, LWG-3724
+struct weird_string_view {
+    const char* data_  = nullptr;
+    unsigned int size_ = 0u;
+
+    weird_string_view()                                    = default;
+    explicit weird_string_view(const weird_string_view&)   = default;
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_all/test.cpp:29]: `copy_constructible` requires implicit copy construction via `convertible_to<const T&, T>`, so this is a move-only view.

File: stl/inc/ranges [hunk 2]
```diff
@@ -1586,8 +1586,8 @@ namespace ranges {
             template <class _Rng>
             _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
                 if constexpr (view<remove_cvref_t<_Rng>>) {
-                    if constexpr (constructible_from<remove_cvref_t<_Rng>, _Rng>) {
-                        return {_St::_View, is_nothrow_constructible_v<remove_cvref_t<_Rng>, _Rng>};
+                    if constexpr (convertible_to<_Rng, remove_cvref_t<_Rng>>) {
```
Comment (CaseyCarter) [inline stl/inc/ranges:1589]: Pre-existing: this shouldn't be constrained regardless of the actual constraint. [range.all.general]/2.1 says that `views::all(E)` is expression-equivalent to "`decay-copy(E)` if the decayed type of `E` models `view`", it notably does not require that `decay-copy(E)` is well-formed.

This distinction didn't matter before adding move-only views to the design: `decay-copy(E)` was always well-formed. After, it is possible that the decayed type of `E` is a move-only view but `decay-copy(E)` is ill-formed because `E` is `const` and/or an lvalue. Nevertheless, we don't want to constrain it away: trying to use an lvalue of move-only view type in view composition _should_ be a hard error reported to the user via `static_assert`. We don't want to silently fall back on non-`view` behavior, we want to make it clear to the user that we understand their intent but that it's not possible to make it so.

TLDR: we should `static_assert` this `convertible_to` requirement in the `_St::_View` branch of `operator()` with a nice error message like "views::all cannot accept an lvalue of a move-only view type".

File: tests/std/tests/P0896R4_views_all/test.cpp [hunk 3]
```diff
@@ -183,4 +210,11 @@ int main() {
         test_one(str);
         assert(ranges::equal(views::all(str), str));
     }
+
+    // Weird views, LWG-3724
+    {
+        weird_string_view v{};
+        static_assert(test_one(v));
+        test_one(v);
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_all/test.cpp:218]: I think this will trigger my suggested `static_assert`. If so, you can remove this test case and or put it in a new "death" test if you are really motivated. (I wouldn't refuse such a test case, but wouldn't require on either.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/ranges (hunk 2 above)
Comment (frederick-vs-ja) [reply stl/inc/ranges:1589]: > TLDR: we should `static_assert` this `convertible_to` requirement in the `_St::_View` branch of `operator()` with a nice error message like "views::all cannot accept an lvalue of a move-only view type".

It seems that the intent of LWG-3724 is to properly constrain the `operator()`, and hence this suggestion is opposite to such intent...

Comment (CaseyCarter) [reply stl/inc/ranges:1589]: Ah, thanks for the reminder: we need to emulate the result of substitution causing the equivalent expression to be ill-formed.

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing so many LWG issue resolutions and improving conformance! :heart_eyes_cat: :heart_eyes_cat: :heart_eyes_cat: :heart_eyes_cat:
